---
title: "UMAP exploration"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## UMAP analysis

First, get the UMAP results as before
```{r load}
library(uwot)

source("./import/import_opc12.R")
source("./import/import_opc90.R")
source("./import/import_opcBulk.R")

source("./functions/opc12_rheg.R")
source("./functions/opc90_rheg.R")

load("./build/opcBulk_DESeq2_results.RData")

umap_bulk <- cbind(data.frame(row.names = bulk$log2$symbol),
                   bulk$log2[,10:ncol(bulk$log2)])

# Join 12 dpi and 90 dpi 10cRNA-seq datasets
umap_opc12 <- cbind(data.frame(row.names = opc12$log2$symbol[opc12$log2$chr != "ERCC"]),
                    opc12$log2[opc12$log2$chr != "ERCC",9 + which(opc12$info$type == "ten-cell")])
umap_opc90 <- cbind(data.frame(row.names = opc12$log2$symbol[opc90$log2$chr != "ERCC"]),
                    opc90$log2[opc90$log2$chr != "ERCC",9 + which(opc90$info$type == "ten-cell")])
umap_10c <- cbind(umap_opc12,umap_opc90)

# Z-score the bulk RNA-seq and 10cRNA-seq datasets separately
umap_bulk_scale <- t(scale(t(umap_bulk)))
umap_10c_scale <- t(scale(t(umap_10c)))

# Join scaled bulk RNA-seq and 10cRNA-seq datasets
umap_all_scale <- cbind(umap_bulk_scale,umap_10c_scale)

# List all genes of interest (bulk DEGs + 10c RHEGs)
DEGs_and_RHEGs <- Reduce(union,list(opc12$candidates$RHEG,
                                    opc90$candidates$RHEG,
                                    bulk$deseq2$de12$genesDE,
                                    bulk$deseq2$de90$genesDE,
                                    bulk$deseq2$de150$genesDE))

# Retain all genes of interest that are not NA (i.e. SD is zero)
umap_all_scale <- umap_all_scale[DEGs_and_RHEGs,]
umap_all_scale <- umap_all_scale[complete.cases(umap_all_scale),]

# Run UMAP
set.seed(0)
embedding_umap <- uwot::umap(X = t(umap_all_scale),n_neighbors = 20)

# Gather information for markers and colors
umap_all.info <- c(paste0("bulk",bulk$info$day,"_",bulk$info$genotype),rep("opc12",sum(opc12$info$type == "ten-cell")),rep("opc90",sum(opc90$info$type == "ten-cell")))
pch_umap <- c(16,1,17,2,15,0,16,16)[as.numeric(factor(umap_all.info))]
col_umap <- c(rep("#000000",6),"#5e3c99","#e66101")[as.numeric(factor(umap_all.info))]
```

Obtain the bulk 12 dpi line (line12) and the orthogonal line (pline) that passes through the median of bulk 150 mutant
```{r lines}
line12x <- c(median(embedding_umap[umap_all.info == "bulk12_WT",1]),median(embedding_umap[umap_all.info == "bulk12_CKO",1]))
line12y <- c(median(embedding_umap[umap_all.info == "bulk12_WT",2]),median(embedding_umap[umap_all.info == "bulk12_CKO",2]))

median12x <- median(embedding_umap[umap_all.info == "bulk12_CKO",1])
median12y <- median(embedding_umap[umap_all.info == "bulk12_CKO",2])

line12m <- diff(line12y) / diff(line12x)
line12b <- median12y - line12m * median12x

a_line12 <- -1*line12m
b_line12 <- 1
c_line12 <- -1*line12b

dist_line12 <- function(x,y) {
  return((a_line12 * x + b_line12 * y + c_line12) / sqrt(a_line12^2 + b_line12^2))
}

median150x <- median(embedding_umap[umap_all.info == "bulk150_CKO",1])
median150y <- median(embedding_umap[umap_all.info == "bulk150_CKO",2])

plinem <- -1 / line12m
plineb <- median150y - plinem * median150x

a_pline <- -1*plinem
b_pline <- 1
c_pline <- -1*plineb

dist_pline <- function(x,y) {
  return(abs(a_pline * x + b_pline * y + c_pline) / sqrt(a_pline^2 + b_pline^2))
}

d <- dist_line12(embedding_umap[,1],embedding_umap[,2]) - dist_pline(embedding_umap[,1],embedding_umap[,2])

```

``` {r }
# Plot UMAP of all datasets
pdf(file = "~/Downloads/UMAP_DEGs_and_RHEGs.pdf",width = 2.25,height = 2.25,pointsize = 7,useDingbats = F)
par(mai = c(0.5,0.5,0,0),mgp = c(1.6,0.6,0))
plot(x = embedding_umap,
     pch = pch_umap,
     col = col_umap,
     lwd = 0.5,
     xlim = c(-3,3),
     ylim = c(-3,3),
     xlab = "UMAP-1",
     ylab = "UMAP-2",
     frame = F,
     axes = F,
     xaxs = "i",
     yaxs = "i")
axis(side = 1,lwd = 0.5,las = 1)
axis(side = 2,lwd = 0.5,las = 1)
legend(x = "topright",legend = c("bulk12-WT","bulk90-WT","bulk150-WT","bulk12-CKO","bulk90-CKO","bulk150-CKO","10c-12","10c-90"),
       pch = c(1,0,2,16,15,17,16,16),
       col = c(rep("#000000",6),"#5e3c99","#e66101"),
       pt.lwd = 0.5,
       box.lwd = 0.5,
       cex = 0.5)
dev.off()
```

Plot the metric as colors
```{r plot, fig.height = 3,fig.width = 3}
library(RColorBrewer)

plot(embedding_umap,col = brewer.pal(n = 9,name = "Reds")[as.numeric(cut(d,breaks = 9))],pch = 16,asp = 1)
lines(line12x,line12y)
abline(plineb,plinem)
text(x = 0.75,y = -0.9,"Bulk 12 dpi vector",adj = c(0,0))
```

Correlate
```{r}
cors <- apply(X = umap_all_scale,MARGIN = 1,FUN = function(x) cor(x = x,y = d,method = "spearman"))

cat(paste(names(sort(x = cors,decreasing = TRUE)[1:20]),sprintf(fmt = "%.3f",as.numeric(sort(x = cors,decreasing = TRUE)[1:20])),sep = "\t"),sep = "\n")

cat(paste(names(sort(x = cors,decreasing = FALSE)[1:20]),sprintf(fmt = "%.3f",as.numeric(sort(x = cors,decreasing = FALSE)[1:20])),sep = "\t"),sep = "\n")


umap_all <- as.matrix(cbind(umap_bulk,umap_10c))
umap_all <- umap_all[DEGs_and_RHEGs,]
umap_all <- umap_all[complete.cases(umap_all),]

umap_all.info <- c(paste0("bulk",bulk$info$day,"_",bulk$info$genotype),rep("opc12",sum(opc12$info$type == "ten-cell")),rep("opc90",sum(opc90$info$type == "ten-cell")))
pch_umap <- c(0,0,2,2,1,1,15,16)[as.numeric(factor(umap_all.info))]
col_umap <- c(rep("#000000",6),"#5e3c99","#e66101")[as.numeric(factor(umap_all.info))]
```

```{r plot, fig.height = 3,fig.width = 3}
plotGeneRank <- 4
plot(x = embedding_umap,
     pch = pch_umap,
     col = brewer.pal(n = 9,name = "Blues")[as.numeric(cut(umap_all[names(sort(x = cors,decreasing = T))[plotGeneRank],],breaks = 9))],
     las = 1,
     xlim = c(-3,3),ylim = c(-3,3),
     main = names(sort(x = cors,decreasing = T))[plotGeneRank])

plot.new()
legend(x = "topleft",legend = levels(factor(umap_all.info))[c(2,1,6,5,4,3,7,8)],pch = c(0,0,2,2,1,1,15,16)[c(2,1,6,5,4,3,7,8)])
```


## Where are the U-group genes in the UMAP?

Get the U-group samples as before
```{r ugroup}
library(pheatmap)
library(RColorBrewer)
library(DESeq2)

f <- "./figures/Figure 4/CIBERSORTx_opc90.txt"
res <- read.table(file = f,header = T,sep = "\t",row.names = 1)

res <- res[,1:(which(names(res) == "P.value") - 1)]

p <- pheatmap(mat = res,
              color = brewer.pal(9,"Reds"),
              clustering_method = "ward.D2",
              border_color = NA,
              lwd = 0.5/0.75,
              treeheight_row = 20,
              treeheight_col = 10,
              fontsize = 7,
              show_rownames = F,
              angle_col = 90,
              silent = T)

o_group <- unlist(as.dendrogram(p$tree_row)[[1]])
n_group <- unlist(as.dendrogram(p$tree_row)[[2]][[1]])
e_group <- c(unlist(as.dendrogram(p$tree_row)[[2]][[2]][[2]][[1]]),unlist(as.dendrogram(p$tree_row)[[2]][[2]][[2]][[2]][[1]]))
u_group <- unlist(as.dendrogram(p$tree_row)[[2]][[2]][[2]][[2]][[2]][[2]])

group <- rep("",56)
group[o_group] <- "O"
group[e_group] <- "E"
group[n_group] <- "N"
group[u_group] <- "U"

annotation <- data.frame(row.names = row.names(res),
                         group = group)
```

G
```{r uumap, fig.height = 3,fig.width = 3}
pch_umap <- c(16,1,16,1,16,1,1,1)[as.numeric(factor(umap_all.info))]
col_umap <- c(rep("#000000",6),"#5e3c99","#e66101")[as.numeric(factor(umap_all.info))]

library(e1071)
dat <- data.frame(embedding_umap[grepl("opc",umap_all.info),],y = factor(c(rep("12 dpi",56),rep("90 dpi",56))))
svmfit <- svm(y ~ .,data = dat,kernel = "linear",scale = F)
beta = drop(t(svmfit$coefs)%*%embedding_umap[grepl("opc",umap_all.info),][svmfit$index,])
beta0 = svmfit$rho



plot(x = embedding_umap[grepl("opc",umap_all.info),],
     pch = pch_umap[grepl("opc",umap_all.info)],
     col = paste0(col_umap[grepl("opc",umap_all.info)],"00"),
     lwd = 0.5/0.75,
     cex = 0.75,
     xlim = c(-3,3),ylim = c(-3,3),
     xlab = "UMAP-1",ylab = "UMAP-2",
     frame = F,axes = F,
     xaxs = "i",yaxs = "i")
points(embedding_umap[c(rep(FALSE,36),rep(FALSE,56),annotation$group == "O"),],pch = 15,col = "#e66101")
points(embedding_umap[c(rep(FALSE,36),rep(FALSE,56),annotation$group == "E"),],pch = 4,col = "#e66101")
points(embedding_umap[c(rep(FALSE,36),rep(FALSE,56),annotation$group == "N"),],pch = 17,col = "#e66101")
points(embedding_umap[c(rep(FALSE,36),rep(FALSE,56),annotation$group == "U"),],pch = 16,col = "#e66101")
abline(beta0 / beta[2], -beta[1] / beta[2],lwd = 0.5/0.75)
axis(side = 1,lwd = 0.5/0.75,las = 1)
axis(side = 2,lwd = 0.5/0.75,las = 1)
text(x = c(-1.5,2),y = c(-1.5,1),labels = c("12 dpi","90 dpi"),cex = 6/7)
legend(x = "topright",legend = c("O","E","N","U"),pch = c(15,4,17,16),col = "#e66101")
```